### PR TITLE
feat: Announcement scheduler

### PR DESCRIPTION
**Pull request by jcbcapps**
# SC-1083

## Proposed changes
Adds ability to schedule an announcement to be published at a future date

---

## Reviewer notes
This is a branch off of the article scheduler branch, which there is a PR for [here](https://github.com/USSF-ORBIT/ussf-portal-cms/pull/258). If you've already given that a glance, then this PR will be easy, cause it's almost entirely the same but for announcements! 

Also please review:
- PR for displaying in the portal: https://github.com/USSF-ORBIT/ussf-portal-client/pull/924
- PR for e2e test: https://github.com/USSF-ORBIT/ussf-portal/pull/194

## Setup
Run everything as usual:
In the CMS: `yarn services:up && yarn dev`
In the portal: `yarn dev`

- Log in to the CMS as a user with the appropriate permissions (admin, manager, etc).
- Create an announcement and set a publish date in the future.
- If you have the `sc-1083/announcement-scheduler` branch checked out in the portal client, navigate to `http://localhost:3000/` and verify that the announcement does not display.

---

### As the original developer, I have

- [ ] Met the acceptance criteria, or will meet them in subsequent PRs or stories
  - ... 
- [ ] Created new stories in Storybook if applicable
  - [ ] Checked that all Storybook accessibility checks are passing
- [ ] Created/modified automated unit tests in Jest
  - [ ] Including jest-axe checks when UI changes
- [ ] Created/modified automated E2E tests in Playwright
  - [ ] Including [axe-playwright](https://github.com/abhinaba-ghosh/axe-playwright) checks when UI changes
  - [ ] Checked that the E2E test build is not failing
- Performed [a11y testing](https://github.com/trussworks/accessibility/blob/master/sample_a11y_testing_process.md):
  - [ ] Checked responsiveness in mobile, tablet, and desktop
  - [ ] Checked keyboard navigability
  - [ ] Tested with [VoiceOver](https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts) in Safari
  - [ ] Checked VO's [rotor menu](https://github.com/trussworks/accessibility/blob/master/README.md#how-to-use-the-rotor-menu) for landmarks, page heading structure and links
  - [ ] Used a browser a11y tool to check for issues (WAVE, axe, ANDI or Accessibility addon tab for Storybook)
- [ ] Requested a design review for user-facing changes
- For any new migrations/schema changes:
  - [ ] Followed guidelines for zero-downtime deploys

### As code reviewer(s), I have

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
  - [ ] Checked that the E2E test build is not failing
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] Considered marking this as accepted even if there are small changes needed
- Performed [a11y testing](https://github.com/trussworks/accessibility/blob/master/sample_a11y_testing_process.md):
  - [ ] Checked responsiveness in mobile, tablet, and desktop
  - [ ] Checked keyboard navigability
  - [ ] Tested with [VoiceOver](https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts) in Safari
  - [ ] Checked VO's [rotor menu](https://github.com/trussworks/accessibility/blob/master/README.md#how-to-use-the-rotor-menu) for landmarks, page heading structure and links
  - [ ] Used a browser a11y tool to check for issues (WAVE, axe, ANDI or Accessibility addon tab for Storybook)

### As a designer reviewer, I have

- [ ] Checked in the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow
- Performed [a11y testing](https://github.com/trussworks/accessibility/blob/master/sample_a11y_testing_process.md):
  - [ ] Checked responsiveness in mobile, tablet, and desktop
  - [ ] Checked keyboard navigability
  - [ ] Tested with [VoiceOver](https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts) in Safari
  - [ ] Checked VO's [rotor menu](https://github.com/trussworks/accessibility/blob/master/README.md#how-to-use-the-rotor-menu) for landmarks, page heading structure and links
  - [ ] Used a browser a11y tool to check for issues (WAVE, axe, ANDI or Accessibility addon tab for Storybook)

---

### CODE REVIEW

File: src/lists/Announcement.ts [hunk 1]
```diff
@@ -94,14 +76,47 @@ const Announcement = list(
...
+        hooks: {
+          resolveInput: async ({ inputData, item, resolvedData }) => {
+            if (
+              inputData.status === ANNOUNCEMENT_STATUSES.PUBLISHED &&
+              item?.status !== ANNOUNCEMENT_STATUSES.PUBLISHED &&
+              !inputData.publishedDate
+            ) {
+              // Set publishedDate if status is being changed to "Published"
+              // only set publishedDate if they didn't set one
+              return DateTime.now().toJSDate()
```
Comment (gidjin) [inline src/lists/Announcement.ts:98]: Suggestion: add tests to test that a date can be set in the future properly and that if one is not set that a time in the near past is set for us. I believe I added examples in the Article tests but let me know if you need a hand with that.

File: src/lists/Announcement.ts [hunk 2]
```diff
@@ -94,14 +76,47 @@ const Announcement = list(
...
+            }
+            return resolvedData.publishedDate
+          },
+          validateInput: async ({
```
Comment (gidjin) [inline src/lists/Announcement.ts:102]: Suggestion: add tests for validation, confirm that a date in the past (beyond 4 hours) calls `addValidationError`, and only on update of publishedDate. I added a similar test for Article, but let me know if you need help with that.

**Review by gidjin — CHANGES_REQUESTED** [review]
LGTM, just needs a couple of unit tests that I mentioned and then I'm happy to approve.

**Comment by jbecker01** [comment]
@jcbcapps when setting a date and time and then saving, it's not updating the status to published. I have to go back and click on publish, which at that point sets the publish time and date to now.

**Comment by jcbcapps** [comment]
> @jcbcapps when setting a date and time and then saving, it's not updating the status to published. I have to go back and click on publish, which at that point sets the publish time and date to now.

@jbecker01 Just pushed up some changes that should fix this!

**Comment by jbecker01** [comment]
:+1:

**Review by gidjin — APPROVED** [review]
LGTM!

**Pull request merged**